File: leetcode_python/539.py
# 539. 最小时间差
# https://leetcode-cn.com/problems/minimum-time-difference/
from typing import List
from collections import defaultdict
from collections import deque
from math import inf
import bisect
import heapq
'''
@File : 539.py
@Time : 2022/01/18 22:12:51
@Author : YuMin Zhang
@State : Indepeedent 
@Thinking :
@Tag : Medium
'''
class Solution:
    def findMinDifference(self, timePoints: List[str]) -> int:
        timePoints.sort(key = lambda x: (x[:2],x[3:]))
        ans = float("inf")
        pre_hour = int(timePoints[0][:2])
        pre_minute = int(timePoints[0][3:])
        for i in range(1, len(timePoints)):
            cur_hour = int(timePoints[i][:2])
            cur_minute = int(timePoints[i][3:])
            sub = (cur_hour - pre_hour) * 60 + cur_minute - pre_minute
            if sub < ans:
                ans = sub
            pre_hour = cur_hour
            pre_minute = cur_minute
        cur_hour = int(timePoints[0][:2])
        cur_minute = int(timePoints[0][3:])
        sub = (24 + cur_hour - pre_hour ) * 60 +  cur_minute - pre_minute
        if sub < ans:
            ans = sub
        return int(ans)

File: leetcode_c_plus_plus/539.cpp
// 539. 最小时间差
// https://leetcode-cn.com/problems/minimum-time-difference/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 539.cpp
@Time : 2022/01/18 22:16:06
@Author : YuMin Zhang
@State : Indepeedent
@Thinking :
@Tag : Medium
**/
class Solution {
public:
    int getMin(string& time) {
        return (int(time[0] - '0') * 10 + int(time[1] - '0')) * 60 + int(time[3] - '0') * 10 + int(time[4] - '0');
    }
    int findMinDifference(vector<string>& timePoints) {
        if(timePoints.size() > 1440) return 0; // 抽屉原理的运用 超过1440 = 24 * 60 必然有一个重复的
        sort(timePoints.begin(), timePoints.end());
        int pre_time = getMin(timePoints[0]);
        int t0_time = pre_time;
        int ans = INT_MAX;
        for(int i = 1; i < timePoints.size(); i++) {
            int cur = getMin(timePoints[i]);
            ans = min(ans, cur - pre_time);
            pre_time = cur;
        }
        ans = min(ans, t0_time + 1440 - pre_time);
        return ans;
    }
};